python/iceberg/api/transforms/projection_util.py

```diff
@@ -18,49 +18,138 @@
+
+
+def truncate_long(name, pred, transform):
+    return truncate_integer(name, pred, transform)
+
+
+def truncate_long_strict(name, pred, transform):
+    return truncate_integer_strict(name, pred, transform)
+
+
+def truncate_decimal(name, pred, transform):
+    boundary = pred.lit.value
+
+    if pred.op == Operation.LT:
+        minus_one = boundary - decimal.Decimal(1)
+        return Expressions.predicate(Operation.LT_EQ, name, transform.apply(minus_one))
+    elif pred.op == Operation.LT_EQ:
+        return Expressions.predicate(Operation.LT_EQ, name, transform.apply(boundary))
+    elif pred.op == Operation.GT:
+        plus_one = boundary + decimal.Decimal(1)
+        return Expressions.predicate(Operation.GT_EQ, name, transform.apply(plus_one))
+    elif pred.op == Operation.GT_EQ:
+        return Expressions.predicate(Operation.GT_EQ, name, transform.apply(boundary))
+    elif pred.op == Operation.EQ:
+        return Expressions.predicate(pred.op, name, transform.apply(boundary))
+
+
+def truncate_decimal_strict(name, pred, transform):
+    boundary = pred.lit.value
+    minus_one = boundary - decimal.Decimal(1)
+    plus_one = boundary + decimal.Decimal(1)
+
+    if pred.op == Operation.LT:
+        return Expressions.predicate(Operation.LT, name, transform.apply(boundary))
+    elif pred.op == Operation.LT_EQ:
+        return Expressions.predicate(Operation.LT, name, transform.apply(plus_one))
+    elif pred.op == Operation.GT:
+        return Expressions.predicate(Operation.GT, name, transform.apply(boundary))
+    elif pred.op == Operation.GT_EQ:
+        return Expressions.predicate(Operation.GT, name, transform.apply(minus_one))
+    elif pred.op == Operation.NOT_EQ:
+        return Expressions.predicate(Operation.NOT_EQ, name, transform.apply(boundary))
+
+
+def truncate_array(name, pred, transform):
+    boundary = pred.lit.value
+
+    if pred.op == Operation.LT or pred.op == Operation.LT_EQ:
+        return Expressions.predicate(Operation.LT_EQ, name, transform.apply(boundary))
+    elif pred.op == Operation.GT or pred.op == Operation.GT_EQ:
+        return Expressions.predicate(Operation.GT_EQ, name, transform.apply(boundary))
+    elif pred.op == Operation.EQ:
+        return Expressions.predicate(pred.op, name, transform.apply(boundary))
+    # todo elif pred.op == Operation.STARTS_WITH:
+
+
+def truncate_array_strict(name, pred, transform):
+    boundary = pred.lit.value
+
+    if pred.op == Operation.LT or pred.op == Operation.LT_EQ:
+        return Expressions.predicate(Operation.LT, name, transform.apply(boundary))
+    elif pred.op == Operation.GT or pred.op == Operation.GT_EQ:
+        return Expressions.predicate(Operation.GT, name, transform.apply(boundary))
+    elif pred.op == Operation.NOT_EQ:
+        return Expressions.predicate(Operation.NOT_EQ, name, transform.apply(boundary))
+
+
+def project_transform_predicate(transform, partiton_name, pred):
+    raise NotImplementedError
```

Comment:
why add a method that isn't used anywhere and only raises an exception?

Reply:
I added it as a placeholder so we track those not implemented functions in python.